lib/client/conntest/ssh.go

```diff
@@ -49,30 +60,212 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
+	}
+
+	hostKeyCallback, err := sshutils.NewHostKeyCallback(sshutils.HostKeyCallbackConfig{
+		GetHostCheckers: func() ([]ssh.PublicKey, error) {
+			return certPublicKeys, nil
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshClientConfig, err := key.ProxyClientSSHConfig(nil, sshNodeAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshClientConfig.HostKeyCallback = hostKeyCallback
+	sshClientConfig.User = req.SSHPrincipal
+
+	dialCtx, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
+
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(dialCtx, "tcp", sshNodeAddr)
+	if err != nil {
+		connDiag, err := s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			uuid.NewString(),
+			types.DiagnosticTraceTypeConnectivity,
+			"Failed to access the host. Please ensure it's network reachable.",
+			err,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+	defer conn.Close()
+
+	_, err = s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		uuid.NewString(),
+		types.DiagnosticTraceTypeConnectivity,
+		"Host is alive and reachable.",
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshConn, sshNewChannel, sshReq, err := ssh.NewClientConn(conn, sshNodeAddr, sshClientConfig)
+	if err != nil {
+		connDiag, err := s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			uuid.NewString(),
+			types.DiagnosticTraceTypeNodeSSHServer,
+			"Failed to open SSH connection. Please ensure Teleport is running.",
+			err,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	_, err = s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		uuid.NewString(),
+		types.DiagnosticTraceTypeNodeSSHServer,
+		"Established an SSH connection.",
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sshClient := ssh.NewClient(sshConn, sshNewChannel, sshReq)
+
+	sshSession, err := sshClient.NewSession()
+	if err != nil {
+		connDiag, err := s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			uuid.NewString(),
+			types.DiagnosticTraceTypeNodeSSHSession,
+			"Failed to create a new SSH Session. Please ensure Teleport is running.",
+			err,
+		))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+	defer sshSession.Close()
+
+	_, err = s.clt.AppendTraceConnectionDiagnostic(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		uuid.NewString(),
+		types.DiagnosticTraceTypeNodeSSHSession,
+		"Created an SSH session.",
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	output, err := sshSession.CombinedOutput("whoami")
```

Comment:
TBH I'm not sure this is necessary, I would probably stop the test after having established the connection.

Reply:
Things changed a little bit but this part is still there.
To call `TeleportClient.SSH` we have two options: interactive or non-interactive mode.
If interactive, we would need to enter the shell and then exit it using Stdin/Stdout
If non-interactive (like we have now), we need to pass a command.

Non-interactive seems simpler.
Interactive means we would need to manage buffers to read and write into the remote process (shell)

As for the actual command in non-interactive mode, I picked `whoami` which seems to be available in any common OS (linux, mac, windows).
Someone can argue that it may not be available, but then we can also argue that to exit a shell we might need to know which shell got spawned and use a different command (do all support the `exist` function?)

Another approach would be to use an SSH escape code to kill the session, but again, seems too complex with little benefit (not sure it would work / exit cleanly).

Reply:
I think it's ok, let's keep it.